src/raqm.c

```diff
@@ -1119,10 +1133,376 @@ _raqm_u32_to_u8_index (raqm_t   *rq,
+      *index = rq->text_len;
+    else
+      *index = 0;
+    return true;
+  }
+
+  RAQM_TEST ("\n");
+
+  for (raqm_run_t *run = rq->runs; run != NULL; run = run->next)
+  {
+    size_t len;
+    hb_glyph_info_t *info;
+    hb_glyph_position_t *position;
+    len = hb_buffer_get_length (run->buffer);
+    info = hb_buffer_get_glyph_infos (run->buffer, NULL);
+    position = hb_buffer_get_glyph_positions (run->buffer, NULL);
+ 
+    for (size_t i = 0; i < len; i++)
+    {
+      delta_x = position[i].x_advance;
+      if (x < (current_x + delta_x))
+      {
+        bool before = false;
+        if (run->direction == HB_DIRECTION_LTR)
+          before = (x < current_x + (delta_x / 2));
+        else
+          before = (x > current_x + (delta_x / 2));
+
+        if (before)
+          *index = info[i].cluster;
+        else
+        {
+          uint32_t curr_cluster = info[i].cluster;
+          uint32_t next_cluster = curr_cluster;
+          if (run->direction == HB_DIRECTION_LTR)
+            for (size_t j = i + 1; j < len && next_cluster == curr_cluster; j++)
+              next_cluster = info[j].cluster;
+          else
+            for (int j = i - 1; i != 0 && j >= 0 && next_cluster == curr_cluster; j--)
+              next_cluster = info[j].cluster;
+
+          if (next_cluster == curr_cluster)
+            next_cluster = run->pos + run->len;
+
+          *index = next_cluster;
+        }
+        if (_raqm_allowed_grapheme_boundary(rq->text[*index],rq->text[*index+1]))
+        {
+          RAQM_TEST ("The start-index is %ld  at position %d \n", *index, x);
+            return true;
+        }
+
+        while (*index < (unsigned)run->pos + run->len)
+        {
+          if (_raqm_allowed_grapheme_boundary(rq->text[*index],rq->text[*index+1]))
+          {
+            *index += 1;
+            break;
+          }
+          *index += 1;
+        }
+        RAQM_TEST ("The start-index is %ld  at position %d \n", *index, x);
+        return true;
+      }
+      else
+        current_x += delta_x;
+    }
+  }
+
+  /* Get rightmost index*/
+  if (rq->resolved_dir == RAQM_DIRECTION_RTL)
+    *index = 0;
+  else
+    *index = rq->text_len;
+
+  RAQM_TEST ("The start-index is %ld  at position %d \n", *index, x);
+
+  return true;
+}
+
+typedef enum
+{
+  RAQM_GRAPHEM_CR,
+  RAQM_GRAPHEM_LF,
+  RAQM_GRAPHEM_CONTROL,
+  RAQM_GRAPHEM_EXTEND,
+  RAQM_GRAPHEM_REGIONAL_INDICATOR,
+  RAQM_GRAPHEM_PREPEND,
+  RAQM_GRAPHEM_SPACING_MARK,
+  RAQM_GRAPHEM_HANGUL_SYLLABLE,
+  RAQM_GRAPHEM_OTHER
+} raqm_grapheme_t;
+
+static raqm_grapheme_t
+_raqm_get_grapheme_break (hb_codepoint_t ch,
+                          hb_unicode_general_category_t category);
+
+static bool
+_raqm_allowed_grapheme_boundary(hb_codepoint_t l_char,
+                                hb_codepoint_t r_char)
+{
+  hb_unicode_general_category_t l_category;
+  hb_unicode_general_category_t r_category;
+  raqm_grapheme_t gb_l_ch, gb_r_ch;
```

Comment:
I think `l_grapheme` and `r_grapheme` would be better names than the cryptic `gb_l_ch`. Or even better `grapheme_l`, `category_l` etc.